File: backend/internal/model/visitante.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Visitante representa o registro de um visitante na granja (livro de registro digital).
type Visitante struct {
	ID                 uuid.UUID  `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	GranjaID           uuid.UUID  `gorm:"type:uuid;not null;index:idx_visitantes_granja" json:"granja_id" validate:"required"`
	Nome               string     `gorm:"type:varchar(200);not null" json:"nome" validate:"required,max=200"`
	Documento          *string    `gorm:"type:varchar(50)" json:"documento,omitempty" validate:"omitempty,max=50"`
	Origem             *string    `gorm:"type:varchar(200)" json:"origem,omitempty" validate:"omitempty,max=200"`
	Motivo             *string    `gorm:"type:varchar(200)" json:"motivo,omitempty" validate:"omitempty,max=200"`
	UltimoContatoAves  *time.Time `gorm:"type:date" json:"ultimo_contato_aves,omitempty"`
	PlacaVeiculo       *string    `gorm:"type:varchar(20)" json:"placa_veiculo,omitempty" validate:"omitempty,max=20"`
	DataEntrada        time.Time  `gorm:"type:timestamp;not null" json:"data_entrada" validate:"required"`
	DataSaida          *time.Time `gorm:"type:timestamp" json:"data_saida,omitempty"`
	Observacao         *string    `gorm:"type:text" json:"observacao,omitempty"`
	CreatedAt          time.Time  `gorm:"not null;default:now()" json:"created_at"`

	// Relacionamentos
	Granja Granja `gorm:"foreignKey:GranjaID" json:"-"`
}

func (Visitante) TableName() string {
	return "visitantes"
}

func (v *Visitante) BeforeCreate(tx *gorm.DB) error {
	if v.ID == uuid.Nil {
		v.ID = uuid.New()
	}
	return nil
}

File: flutter_app/lib/features/sanidade/domain/entities/visitante.dart
/// Registro de visitante em uma granja (livro de registro digital).
///
/// Matches backend DTO VisitanteResponse:
/// - granja_id (not lote_id)
/// - data_entrada (not data_visita)
/// - documento, origem, ultimo_contato_aves, placa_veiculo
/// - data_saida
/// Endpoint: GET /granjas/:id/visitantes (not /lotes/:id/visitantes)
class Visitante {
  final String id;
  final String granjaId;
  final String nome;
  final String? documento;
  final String? origem;
  final String? motivo;
  final String? ultimoContatoAves;
  final String? placaVeiculo;
  final String dataEntrada;
  final String? dataSaida;
  final String? observacao;
  final String createdAt;

  const Visitante({
    required this.id,
    required this.granjaId,
    required this.nome,
    this.documento,
    this.origem,
    this.motivo,
    this.ultimoContatoAves,
    this.placaVeiculo,
    required this.dataEntrada,
    this.dataSaida,
    this.observacao,
    required this.createdAt,
  });

  factory Visitante.fromJson(Map<String, dynamic> json) {
    return Visitante(
      id: json['id'] as String,
      granjaId: json['granja_id'] as String,
      nome: json['nome'] as String,
      documento: json['documento'] as String?,
      origem: json['origem'] as String?,
      motivo: json['motivo'] as String?,
      ultimoContatoAves: json['ultimo_contato_aves'] as String?,
      placaVeiculo: json['placa_veiculo'] as String?,
      dataEntrada: json['data_entrada'] as String,
      dataSaida: json['data_saida'] as String?,
      observacao: json['observacao'] as String?,
      createdAt: json['created_at'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'granja_id': granjaId,
      'nome': nome,
      if (documento != null) 'documento': documento,
      if (origem != null) 'origem': origem,
      if (motivo != null) 'motivo': motivo,
      if (ultimoContatoAves != null) 'ultimo_contato_aves': ultimoContatoAves,
      if (placaVeiculo != null) 'placa_veiculo': placaVeiculo,
      'data_entrada': dataEntrada,
      if (dataSaida != null) 'data_saida': dataSaida,
      if (observacao != null) 'observacao': observacao,
      'created_at': createdAt,
    };
  }
}
